docs/src/arrays.md

```diff
@@ -0,0 +1,219 @@
+
+```julia
+function rrule(::typeof(*), A::Matrix, B::Matrix)
+    function times_pullback(ΔC)
+        ∂A = @thunk(ΔC * B')
+        ∂B = @thunk(A' * ΔC)
+        return (NO_FIELDS, ∂A, ∂B)
+    end
+    return A * B, times_pullback
+end
+```
+
+### Matrix inversion
+
+$$C A = A C = I$$
+
+Again, we derived the differential identity as
+
+$$dC = -C dA C$$
+
+Multiplying by $\overline{C}^H$ and taking the trace,
+
+$$ds = \Re\left(\operatorname{tr}(\overline{C}^H dC)\right) = \Re\left(\operatorname{tr}(-\overline{C}^H C dA C)\right)$$
+
+Applying the trace identities to manipulate into the desired form,
+
+$$ds = \Re\left(\operatorname{tr}(-C \overline{C}^H C dA)\right),$$
+
+we can now solve for $\overline{A}$:
+
+$$\overline{A} = (-C \overline{C}^H C)^H = -C^H \overline{C} C^H$$
+
+We can implement the resulting `rrule` as:
+
+```julia
+function rrule(::typeof(inv), A::Matrix)
+    C = inv(A)
+    function inv_pullback(ΔC)
+        ∂A = -C' * ΔC * C'
+        return (NO_FIELDS, ∂A)
+    end
+    return C, inv_pullback
+end
+```
+
+## More examples
+
+For more instructive examples of matrix and vector rules, see [Giles' paper](https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf) and the [LinearAlgebra rules in ChainRules](https://github.com/JuliaDiff/ChainRules.jl/tree/master/src/rulesets/LinearAlgebra).
+
+## Generalizing to multidimensional arrays
+
+For both forward- and reverse-mode rules, the first step was to write down the differential identities, which followed directly from the scalar differential identities.
+This approach follows for arrays, but it's easier to work in component form.
+Consider the function `f(A) = sum(abs2, A::Array{T,3}; dims=2)`, which we write as
+
+$$C_{i1k} = \sum_{j} |A_{ijk}|^2 = \sum_{j}\Re(A_{ijk}^* A_{ijk})$$
+
+The differential identity is
+
+$$dC_{i1k} = \sum_j \Re(dA_{ijk}^* A_{ijk} + A_{ijk}^* dA_{ijk}) \\
+= \sum_j \Re((A_{ijk}^* dA_{ijk})^* + A_{ijk}^* dA_{ijk})\\
+= \sum_j 2 \Re(A_{ijk}^* dA_{ijk}),$$
+
+where in the last step we have used the identity that the sum of its number and its conjugate is twice the real part of the number.
+
+We can then derive the reverse-mode rule.
+The array form of the desired identity will be
+
+$$ds = \Re \left(\sum_{ik} \overline{C}_{i1k}^* dC_{i1k}\right) = \Re \left(\sum_{ijk} \overline{A}_{ijk}^* dA_{ijk} \right)$$
+
+We can plug in the differential identity into the middle expression to get
+
+$$ds = \Re \left(\sum_{ijk} 2 \overline{C}_{i1k}^* \Re(A_{ijk}^* dA_{ijk})\right) = \Re \left(\sum_{ijk} 2 \Re(\overline{C}_{i1k}) A_{ijk}^* dA_{ijk}\right).$$
+
+We can now solve for $\overline{A}$:
+
+$$\overline{A}_{ijk} = (2 \Re(\overline{C}_{i1k}) A_{ijk}^*)^* = A_{ijk} \cdot 2\Re(\overline{C}_{i1k})$$
+
+Because none of this derivation really depended on the index (or indices), we can easily implement the `rrule` more generically using broadcasting:
+
+```julia
+function rrule(::typeof(sum), ::typeof(abs2), A::Array; dims = :)
```

Comment:
```suggestion
function rrule(::typeof(sum), ::typeof(abs2), A::Array{ComplexF64}; dims = :)
```

Reply:
These rules should work for any `Real` and `Complex`, so any objection to `Union{Real,Complex}`?